server/sonar-server/src/main/java/org/sonar/server/permission/index/AuthorizationIndexer.java

```diff
@@ -0,0 +1,202 @@
+  private final EsClient esClient;
+
+  public AuthorizationIndexer(DbClient dbClient, EsClient esClient) {
+    this.executor = new ThreadPoolExecutor(0, 1, 0L, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+    this.dbClient = dbClient;
+    this.esClient = esClient;
+  }
+
+  /**
+   * Index issues authorization and project measures authorization indexes only when they are empty
+   */
+  public void indexAllIfEmpty() {
+    Future submit = executor.submit(() -> {
+      if (isIndexEmpty(IssueIndexDefinition.INDEX, IssueIndexDefinition.TYPE_AUTHORIZATION) ||
+        isIndexEmpty(ProjectMeasuresIndexDefinition.TYPE_PROJECT_MEASURES, ProjectMeasuresIndexDefinition.TYPE_AUTHORIZATION)) {
+        truncate(IssueIndexDefinition.INDEX, IssueIndexDefinition.TYPE_AUTHORIZATION);
+        truncate(ProjectMeasuresIndexDefinition.TYPE_PROJECT_MEASURES, ProjectMeasuresIndexDefinition.TYPE_AUTHORIZATION);
+        try (DbSession dbSession = dbClient.openSession(false)) {
+          index(new AuthorizationDao().selectAll(dbClient, dbSession));
+        }
+      }
+    });
+    try {
+      Uninterruptibles.getUninterruptibly(submit);
+    } catch (ExecutionException e) {
+      Throwables.propagate(e);
+    }
+  }
+
+  private boolean isIndexEmpty(String index, String type) {
+    SearchResponse issuesAuthorizationResponse = esClient.prepareSearch(index).setTypes(type).setSize(0).get();
+    return issuesAuthorizationResponse.getHits().getTotalHits() == 0;
+  }
+
+  private void truncate(String index, String type) {
+    BulkIndexer.delete(esClient, index, esClient.prepareSearch(index).setTypes(type).setQuery(matchAllQuery()));
+  }
+
+  public void index(List<String> projectUuids) {
+    checkArgument(!projectUuids.isEmpty(), "ProjectUuids cannot be empty");
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      AuthorizationDao dao = new AuthorizationDao();
+      index(dao.selectByProjects(dbClient, dbSession, projectUuids));
+    }
+  }
+
+  private void index(Collection<AuthorizationDao.Dto> authorizations) {
+    if (authorizations.isEmpty()) {
+      return;
+    }
+    int count = 0;
+    BulkRequestBuilder bulkRequest = esClient.prepareBulk().setRefresh(false);
+    for (AuthorizationDao.Dto dto : authorizations) {
+      bulkRequest.add(newIssuesAuthorizationIndexRequest(dto));
+      bulkRequest.add(newProjectMeasuresAuthorizationIndexRequest(dto));
+      count++;
+      if (count >= MAX_BATCH_SIZE) {
+        EsUtils.executeBulkRequest(bulkRequest, BULK_ERROR_MESSAGE);
+        bulkRequest = esClient.prepareBulk().setRefresh(false);
+        count = 0;
+      }
+    }
+    EsUtils.executeBulkRequest(bulkRequest, BULK_ERROR_MESSAGE);
+    esClient.prepareRefresh(IssueIndexDefinition.INDEX).get();
+    esClient.prepareRefresh(ProjectMeasuresIndexDefinition.TYPE_PROJECT_MEASURES).get();
+  }
+
+  public void index(String projectUuid) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
```

Comment:
dbSession should not be opened here but passed as a parameter.
Ideally sessions are opened only by web services then all backend gets it through parameters.